src/session/components/message_entry.rs

```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
         self.imp().text_view.buffer().set_text(&text);
     }
 
+    /* formatted_text is being replaced every time text buffer has changed,
+     * so it's very unefficient to parse markdown text when it's not nesessary,
+     * instead we should use a separate method, that will be called only when composing a message */
+
+    pub(crate) async fn parsed_formatted_text(&self) -> Option<FormattedText> {
+        let text = self
+            .imp()
+            .formatted_text
+            .borrow()
+            .clone()
+            .map(|f| f.0)
+            .unwrap_or_default();
+
+        let chat = self.imp().chat.borrow().clone();
+        let client_id = chat.unwrap().session().client_id();
+        functions::parse_markdown(text, client_id)
+            .await
+            .map(|text| {
+                let EnumFormattedText::FormattedText(text) = text;
+                text
+            })
+            .ok()
+    }
+
+    pub(crate) fn chat(&self) -> Option<Chat> {
+        self.imp().chat.borrow().clone()
+    }
+
+    pub(crate) fn set_chat(&self, chat: Option<Chat>) {
+        if self.chat() != chat && chat.is_some() {
+            self.imp().chat.replace(chat);
+        }
```

Comment:
1. I would keep the scheme/style we use to set properties
2. Testing for `Some` is pointless
3. You forgot to notify the property

```suggestion
        if self.chat().as_ref() == chat {
            return;
        }
        self.imp().chat.set(chat);
        self.notify("chat");
```